### PR TITLE
feat: implement worker queue backed using surrealdb

### PR DESCRIPTION
**Pull request by danigiri**
We implement surrealdb support for broccoli.

## main features
- surrealdb support, lazily creating the necessary tables on a per-topic-basis, which helps in debugging, diagnostics, and performance
- feature flags added to control dependencies
- main ordered queue table is a timeseries, and all operations should use indexes, O(k), including cancellation
- support live queries for added performance, but at the broker implementation level, the high-level code still does polling
- can have multiple consumers on the same topic
  - multiple tasks or instances of consumer processes will be protected by the database transaction model following an optimistic locking model provided by surrealdb
- tests of implemented features pass, and added additional
- throw unsupported on ttl
- priority support
- performance benchmarking added

## proposed API changes
- surrealdb performs much better in batch mode or in live mode, and individual record fetching, so proposing to add:

```
    /// Attempts to consume up to a number of messages from the specified queue.
    /// Does not block if not enough messages are available, and returns immediately.
    ///
    /// # Arguments
    /// * `queue_name` - The name of the queue.
    /// * `batch_size` - Maxium number of messages to try to consume.
    /// # Returns
    /// A `Result` containing an `Vec(String)` with the available message(s)
    /// and a `BroccoliError` on failure.
    async fn try_consume_batch
```
Backwards compatibility is provided by a default trait implementation, which the surrealdb implementation overrides for 10x performance gains.

## proposed config changes
- Added a consumer configuration  called `handler_ack`.  If we use `process_message`or `process_messages_with_handlers` we used to always ack messages. This prevents use cases like:
  - having the handler spawn a long-running task and returning immediately
  - having some other type of reaper process that is untied
  If this is set to false, successful Ok()'s from the callback fn's do not trigger an acknowledge and it's the caller's responsibility to ACK as needed. Note that handler errors still always trigger a reject.
  Default is `true`, maintaining strict backwards compatibility

- Added a consumer config called `consume_wait`. This sets on how long to wait in tight consumer loops, which allows functions running those loops to be stopped in a `tokio::spawn` thread. Defaults to zero to lower performance impact while still achieving the same effect.

## todo
- performance benchmarking still needs some work
- docker tests
- add more polish: more retries for transient transactions and move all mutations to `BEGIN TRANSACTION` even if they are not manifesting RTN
- add more cleanup checks on tests, namely, verify queues are empty, etc.
- handler callbacks tests flakey (only) when using in-memory, though persistent surrealdb is 100% fine

## out of scope / future
- add patch to have full live query support as opposed to current top-level polling (now it's mixed mode)
- ttl support
- fairness support
- only tested on surrealdb 2.3.4

### CODE REVIEW

**Comment by densumesh** [comment]
Thank you for contributing @danigiri! I will review this PR when I get a chance this week! Also aside, why do you think that surrealdb is a good queuing solution? Just genuinely curious here

**Comment by danigiri** [comment]
> Also aside, why do you think that surrealdb is a good queuing solution? Just genuinely curious here

No problem, it's a good question. The main reasons would be:
- that it's a good match if you are already using surrealdb and do not want to add another moving piece of infra
- timeseries-like feature in surrealdb fits reasonably well and is relatively performant to a simple time-ordered queue
- surrealdb has a live query feature which we want to exploit that also fits well, namely you can keep a connection open and receive pushes from the server, which means no polling and less roundtrips/connections to the DB

At some scale point a dedicated (rabbitMQ, kafka, etc.) infra would probably make sense, but surrealdb would do well in medium-scale scenarios

**Comment by densumesh** [comment]
Can you run `cargo clippy --features surrealdb --fix --allow-dirty` to fix all of the clippy lints. I know I have some annoying lints turned on but most of them are pretty helpful and good for maintaining readability

File: src/brokers/surrealdb/broker.rs [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+pub(crate) struct InternalSurrealDBBrokerQueuedMessage {
+    pub(crate) message_id: RecordId, // this is the message id: queue_name:task_id
+    pub(crate) priority: i64,        // message priority copy, to use for sorting in consumption
+}
+
+#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
+pub(crate) struct InternalSurrealDBBrokerProcessingMessage {
+    pub(crate) message_id: RecordId, // this is the message id: queue_name:task_id
+    pub(crate) priority: i64,        // if we reenqueue, we set the same priority
+}
```
Comment (densumesh) [inline src/brokers/surrealdb/broker.rs]: nit: Do we need two separate structs here?

File: src/brokers/surrealdb/broker.rs [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+pub(crate) struct InternalSurrealDBBrokerFailedMessage {
+    pub(crate) original_msg: InternalSurrealDBBrokerMessage, // full original message
+}
+
+// consumer lock
+lazy_static! {
+    // TODO: to improve performance we could have a hash table of mutexes, or sharding
+    static ref consume_mutex: std::sync::Arc<tokio::sync::Mutex<u8>> =
+        std::sync::Arc::new(tokio::sync::Mutex::new(0));
+}
```
Comment (densumesh) [inline src/brokers/surrealdb/broker.rs]: is this bc of lack of atomicity in surrealdb? Seems to me like this is here to ensure that reads are always consistent. Is there no other way to ensure atomicity within surrealdb?

File: src/brokers/surrealdb/broker.rs [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+            .select(queue_table)
+            .range(
+                vec![Value::default(), Value::default()] // note default is 'None'
+                ..vec![Value::from_str("time::now()").unwrap_or(Value::default()),Value::default()],
+            ) // should notify when future becomes present
+            .live()
+            .await
+            .map_err(|err| BroccoliError::Broker(format!("Could not consume: {:?}", err)))?;
+        let mut queued_message: Result<InternalSurrealDBBrokerQueuedMessageRecord, BroccoliError> =
+            Err(BroccoliError::NotImplemented);
```
Comment (densumesh) [inline src/brokers/surrealdb/broker.rs:264]: I dont think this is the right error to use here

File: src/brokers/surrealdb/broker.rs [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+    }
+
+    /// Consumes a message from the specified queue, blocking until a message is available.
+    /// Uses live querying, so if there are no messages yet, it will block efficiently without polling
+    /// # Arguments
+    /// * `queue_name` - The name of the queue.
+    ///
+    /// # Returns
+    /// A `Result` containing the message as a `String`, or a `BroccoliError` on failure.
+    async fn consume(
```
Comment (densumesh) [inline src/brokers/surrealdb/broker.rs:237]: why do we not need a mutex lock here and we do in `try_consume`?

File: src/brokers/surrealdb/utils.rs [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+/// this is the failed messages table
+fn failed_table(queue_name: &str) -> String {
+    format!("{}___failed", queue_name)
+}
+
+/// this is an index to go from <queue_name>___index:[<taskid>,<queuename>] to the queue table
+/// in O(k) time
+fn index_table(queue_name: &str) -> String {
+    format!("{}___index", queue_name)
+}
```
Comment (densumesh) [inline src/brokers/surrealdb/utils.rs]: Am i correct that this is a lookup table to see the queue a specific task id is in? Do we need this if we ask the user to specify the queue name anyway

File: src/brokers/surrealdb/utils.rs [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+    queue_name: &str,
+    task_id: &String,
+    priority: i64,
+    when: String,
+    err_msg: &'static str,
+) -> Result<(), BroccoliError> {
+    let queue_table = self::queue_table(queue_name);
+    // TODO: look at building the record programmatically, move when to typed
+    // compromise here is that we do explicit casting of the uuid, if it's not correct it will fail
+    let queue_record_str = format!("{}:[{},<uuid>'{}']", queue_table, when, task_id);
```
Comment (densumesh) [inline src/brokers/surrealdb/utils.rs]: I think having some functions on the RecordId struct to build these strings will make this alot easier to read and makes it cleaner.

File: src/brokers/surrealdb/utils.rs [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+// we add an entry into a queue index, index:[messageid,queue_name] {queue_id} where
+// queue_id is basically: queue:[timestamp,messageid].
+// We can use this index we can do cancellations in O(k) time and workaround parser issues
+async fn add_to_queue_index(
+    db: &Surreal<Any>,
+    queue_name: &str,
+    task_id: &String,
+    queue_id: RecordId, // queue:[timestamp, task_id]
+    err_msg: &'static str,
+) -> Result<(), BroccoliError> {
```
Comment (densumesh) [inline src/brokers/surrealdb/utils.rs]: Does this index just contain the task ids for a specific queue? or is it a join table between the queue table and the actual tasks themselves?

**Review by densumesh — CHANGES_REQUESTED** [review]
Left some comments some are nits but responses would be appreciated!

**Comment by danigiri** [comment]
Thanks very much for the review, when I have a time slot I will do:
a) `cargo clippy --features surrealdb --fix --allow-dirty` no problem
b) will look at your comments and try to address them one by one

File: src/brokers/surrealdb/broker.rs (hunk 1 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: Good question, rationale was to have type safety and explicitly distinguish between processing and queued messages, any suggestions on a meaningful name that represents both?

Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: It's a semantic distinction at the moment

File: src/brokers/surrealdb/broker.rs (hunk 2 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: Surrealdb has transactions but had issues with building the queries on the first try.

After some further experimentation there is a way, I have a solution in mind for an upcoming commit. Did not want to delay a review too much and have a massive set of changes. This works as long as you shard consumer instances per topic and should perform. Looks like we have two options:
a) this is OK to accept until commit for atomicity is moved to surrealdb
b) wait until commit is added that adds atomicity support in database

File: src/brokers/surrealdb/broker.rs (hunk 3 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:264]: The idea there is that TTL is not implemented (nor easily implementable) in surrealdb so doing a fail fast, would a `Publish` be a better choice?

File: src/brokers/surrealdb/broker.rs (hunk 4 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:237]: Good catch, the blocking live consumer should have equivalent behaviour with locking with future messages. Unfortunately we cannot do live queries with transactions due to a SurrealDB SDK limitation.

The PR is still functional in the sense that try_consume is thread-safe while two live consumer threads are not. Will give it some thought to see if there is a solution.

Quick thought:
1) `try_consume` will be thread safe across multiple instances with database transactions
2) `consume` will be thread safe only within an instance using a lock, requiring topic sharding if we have multiple consumer instances, which is not ideal, but not terrible either

Once surrealdb fixes the SDK limitation, `consume` can quickly be moved to full thread safety across instances

File: src/brokers/surrealdb/utils.rs (hunk 5 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: yes, it's a lookup table.
the extra `___index` table is needed as the queue table is ordered by time and by-id accesses require a full scan, unless we add a field index, so there are three options:

a) when connecting, check if there is a field index and add one: consumer or producer require index creation permissions, need to an extra query every time we connect or create a topic
b) create this table
c) ask admins to add the index manually once a topic is created (error-prone and manual), or add an equivalent trigger

b) is easy and fulfills the philosophy of lazy create if not exists (like in redis, etc.), less permissions surface, and no admin work

File: src/brokers/surrealdb/utils.rs (hunk 6 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: Agreed, but did not find an official way to build this programmatically, and the `create` builder chain does not accept a range either.

Drawing inspiration from the surrealdb programmatic range build expression`type::thing($queue_table,type::range([[None,None],[time::now(),None]]))` will see if there is an equivalent in the rust SDK, otherwise can move to a helper function. In the latter case, once the SDK allows for this, we can quickly replace it.

At least this is type safe as the formatted string is used in the `RecordId::from_str` and the output of that is used as parameter in the `create` builder chain, so there is no dynamic query build with potential injection.

File: src/brokers/surrealdb/utils.rs (hunk 7 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: just the ids, haven't tried but when doing manual exploration of the data one can leverage features like https://surrealdb.com/blog/beyond-sql-joins-exploring-surrealdbs-multi-model-relationships to do joins basically

File: src/brokers/surrealdb/utils.rs (hunk 6 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: Have been trying to coerce an `uuid` type into a surrealdb `Value` but have not been able to unfortunately, will leave this parse as-is and move it to a function for future refactor

**Comment by danigiri** [comment]
Ran clippy and the issues I introduced, the rest I think were there already, this caught my attention:

> warning: item name ends with its containing module's name  

we could fix it  by renaming stuff or `allowed-prefixes = `

File: src/brokers/surrealdb/broker.rs (hunk 1 above)
Comment (densumesh) [reply src/brokers/surrealdb/broker.rs]: Hmm, I think just keeping it as a `InternalSurrealDBBrokerMessage` is fine, because its the same message that gets passed around, regardless of source

File: src/brokers/surrealdb/broker.rs (hunk 2 above)
Comment (densumesh) [reply src/brokers/surrealdb/broker.rs]: I think we can merge for now, but just flagged so i could understand

File: src/brokers/surrealdb/broker.rs (hunk 3 above)
Comment (densumesh) [reply src/brokers/surrealdb/broker.rs:264]: Is this functionality for TTL? I thought it was for consuming messages

File: src/brokers/surrealdb/broker.rs (hunk 4 above)
Comment (densumesh) [reply src/brokers/surrealdb/broker.rs:237]: Thats interesting, so for live consumers, we can't have another thread that is also a live consumer on the surrealdv instance? Does this mean that the call the surreal will fail if we have another thread that tries to acquire a live consumer? I think we should fail prior to the surreal call with some atomic bool that is set if another thread is consuming.

File: src/brokers/surrealdb/utils.rs (hunk 5 above)
Comment (densumesh) [reply src/brokers/surrealdb/utils.rs]: So does this allow for the ability to get the queue_name for each task_id? Or does this give us the index in the queue_table to get that task?

File: src/brokers/surrealdb/utils.rs (hunk 6 above)
Comment (densumesh) [reply src/brokers/surrealdb/utils.rs]: sounds good, as long as its type safe it should be fine

**Comment by densumesh** [comment]
> Ran clippy and the issues I introduced, the rest I think were there already, this caught my attention:
> 
> >

this is fine just want to get all the autofixable ones doen

File: src/brokers/surrealdb/broker.rs (hunk 1 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: Was having a thought at this, made a proposal with a type that holds both use cases, named `InternalSurrealDBBrokerdMessageEntry`, could reuse `InternalSurrealDBBrokerMessage` but that includes the full payload which may be potentially large and we don't need it for the auxiliary tables. Let me know what you think

File: src/brokers/surrealdb/broker.rs (hunk 2 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: agreed to merge, it's not ideal but definitely usable, and can check it out in a future PR

File: src/brokers/surrealdb/broker.rs (hunk 3 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:264]: Not sure if I understand the question, I will rephrase.

Thea idea is that if we detect that the user has requested the TTL feature in the publish options we fail fast and return that the feature is not implemented in surrealdb. 
It's not easy to implement as surrealdb has not concept of timers, there are some ideas on how to do it but they are best left for the future.

I can remove the check and let the user publish, but it could be misleading as the user would expect messages to autoexpire (and they won't).

Could add a clarifying comment

File: src/brokers/surrealdb/broker.rs (hunk 4 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:237]: Added the mutex to the live consume as well. Recap:

Within a single consumer process instance:
  - Consumption with `try_consume` is thread safe, and consumers will just block until the lock becomes available
  - Fair FIFO is guaranteed by tokio mutex (see https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html), namely "Tokio’s Mutex operates on a guaranteed FIFO basis. This means that the order in which tasks call the [lock](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#method.lock) method is the exact order in which they will acquire the lock."
  - Even if multiple threads try to read from the topic we are OK
  - live `consume` is also safe but can have multiple reads if multiple threads consume from the same topic, where one client thread will succeed and the others will fail with an error from `remove_from_queue`
     - we can't add the lock before the live consume as we'll block everyone
     - multiple live consumers can be OK if they watch for this error and recover from it, with a fastest-consumer-wins model
  - live `consume` is completely fine if we have only one thread per topic

With multiple consumer process instances:
- safety is only guaranteed if we don't have multiple processes reading from the same topic

Given the above model I'd say we can go on, as it's quite useful, and add transactions in the future for multiple consumer processes. A multi-threaded first-consumer-wins with per-topic sharding mutliple process implementation model will perform well in many cases, if needing more performance,one should probably move to a dedicated queue infrastructure (as opposed to a generic DB like surrealdb).

WDYT?

File: src/brokers/surrealdb/utils.rs (hunk 5 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: The second case, given a queue name, we get the table that holds the index.
Once we have the table, with the oroginal task id and queue name we can do:
```
`index_table`:[<uuid>`task_id`, `queue_name`] --> `queue`:[timestamp, `task_id`]
```

which means we can operate on the internal queue entry it without a table scan using the timestamp (for cancellations, removals, ...)

Let me know if unclear

File: src/brokers/surrealdb/utils.rs (hunk 7 above)
Comment (danigiri) [reply src/brokers/surrealdb/utils.rs]: See comment above, used for this mapping

```
`index_table`:[<uuid>`task_id`, `queue_name`] --> `queue`:[timestamp, `task_id`]
```

**Comment by danigiri** [comment]
Just recently completed a project and can look at this. Thanks again for the review and comments.

- Addressed most if not all comments 
- made a small implementation change following up on the comments on the live consumption
- ran clippy for autofixable stuff (looks like it changed a couple of imports)

Next:
- I am getting a weird compile error on one of the tests that is triggered in `cargo test` but not within vscode, want to see the GH actions output as well to help diagnose
- address any further comments or follow ups
- think if there is a way to test the live consumer further (low prio)

**Comment by danigiri** [comment]
When evaluating concurrency, looks like surrealdb is <10% of the times returning recently deleted data, even when using a single thread and a local lock

File: src/brokers/surrealdb/broker.rs (hunk 1 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs]: Unified in latest PR

File: src/brokers/surrealdb/broker.rs (hunk 4 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:237]: Added a mutex and a transaction model, stronger safety in-process but still guarantees across processes

**Comment by danigiri** [comment]
Tried to address all of your comments above
Also added transactions and support for multiple readers, both intra and inter-process.
In the future could look at removing the in-process mutex but for now this looks fully functional and first-consumer-wins.

so looks good to re-review @densumesh , thanks in advance

```
     Running tests/edge_cases.rs (target/debug/deps/edge_cases-e38da489b8692b3f)

running 8 tests
test test_invalid_broker_url ... ok
test test_ttl_not_implemented ... ok
test test_empty_payload ... ok
test test_concurrent_consume ... ok
test test_very_large_payload ... ok
test test_process_messages ... ok
test test_process_messages_with_handlers ... ok
test test_message_ordering ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.21s

     Running tests/fairness.rs (target/debug/deps/fairness-4140169bda2ddffe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/happy_path.rs (target/debug/deps/happy_path-54349c6ba58d2c1b)

running 9 tests
test test_publish_and_consume ... ok
test test_message_cancellation ... ok
test test_message_auto_ack ... ok
test test_message_acknowledgment ... ok
test test_batch_publish_and_consume ... ok
test test_message_retry ... ok
test test_message_priority ... ok
test test_delayed_message ... ok
test test_scheduled_message ... ok

test result: ok. 9 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 3.13s

   Doc-tests broccoli_queue

running 2 tests
test src/queue.rs - queue::BroccoliQueue::process_messages_with_handlers (line 770) - compile ... ok
test src/queue.rs - queue::BroccoliQueue::process_messages (line 620) - compile ... ok
```

**Comment by danigiri** [comment]
looking at perf tests now with criterion

**Comment by danigiri** [comment]
Latest commit adds a benchmark for surrealdb, on top of that:
- fixed typo
- added missing include
- removed leftover diagnostics data

Run benchmark
```
RUST_TEST_THREADS=1 cargo bench --features surrealdb  --color always --verbose --bench surrealdb_benchmark                                                     
```

Totally reviewable

The test url used in the benchmark is detected as a secret, which is a false positive 🤷‍♂️, will use an env var to get rid of it in the next commit

**Comment by danigiri** [comment]
Removed hardcoded benchmark url, now settable using an env var

File: src/error.rs [hunk 8]
```diff
@@ -72,6 +79,13 @@ pub enum BroccoliError {
     #[error("Redis error: {0}")]
     Redis(#[from] redis::RedisError),
 
+    /// Represents Redis-specific errors.
+    ///
+    /// This variant wraps the underlying Redis error.
```
Comment (densumesh) [inline src/error.rs]:
```suggestion
    /// Represents SurrealDB-specific errors.
    ///
    /// This variant wraps the underlying SurrealDB error.
```

File: src/error.rs [hunk 9]
```diff
@@ -12,6 +12,13 @@ pub enum BroccoliError {
     #[error("Broker error: {0}")]
     Broker(String),
 
+    /// Errors that occur during broker implementations related to concurrent operations not being idempotent
+    ///
+    /// # Examples
+    /// - Multiple concurrent reads on the same topic
```
Comment (densumesh) [inline src/error.rs:19]: Should clarify in the doc that this is only for SurrealDB, as the other brokers do support concurrent reads

File: src/queue.rs [hunk 10]
```diff
@@ -238,12 +246,20 @@ impl ConsumeOptionsBuilder {
         self
     }
 
+    /// Sets whether to consume from a fairness queue.
```
Comment (densumesh) [inline src/queue.rs]: Better doc comment here. Keep in mind these comments get used to generate the rustdoc

File: src/queue.rs [hunk 11]
```diff
@@ -706,6 +729,10 @@ impl BroccoliQueue {
                     future_handles.push(handle);
                 }
             } else {
+                let tid = tokio::task::id();
+                eprintln!("[{}] About to consume", &tid);
```
Comment (densumesh) [inline src/queue.rs]: Should probably use `log::debug!` for this or remove entirely

File: tests/edge_cases.rs [hunk 12]
```diff
@@ -184,55 +190,182 @@ async fn test_concurrent_consume() {
...
-                !task_exists,
-                "Message should be cleaned up after acknowledgment"
-            );
+async fn error_handler(_: TestMessage, err: BroccoliError) -> Result<(), BroccoliError> {
+    // helper function to test process_messages_with_handlers
+    panic!("Should not invoke the error handler in testing {}", err);
+}
+
+#[tokio::test]
+async fn test_process_messages() {
```
Comment (densumesh) [inline tests/edge_cases.rs]: should prob move this test and `test_process_messages_with_handlers` to the happy path file as these don't test edge cases but expected use cases

**Review by densumesh — CHANGES_REQUESTED** [review]
All in all looks great! just a couple of minor nits and I think we are good to go!

**Comment by densumesh** [comment]
You may also need to rebase onto main as a couple of changes have been made in the meantime

File: src/queue.rs (hunk 10 above)
Comment (danigiri) [reply src/queue.rs]: rewritten, feedback welcome

**Comment by danigiri** [comment]
many thanks for the review, welcome your thoughts and comments

- addressed your comments
- added configurable zero waits in some parts of the tight loops
- let me know what you think on the reword for the configuration value, more than happy to accept a better wording TBH
- merged latest code so I see some of the recent changes like more logging (I think I did it correctly, let me know otherwise) 
- benchmark outcome

Screenshot 2025-05-20 at 16 34 20

so none of the thousands of iterations in the benchmark failed which is a good sign, but this hasn't been tested besides the automated tests include in the code

**Comment by danigiri** [comment]
Made another change, context:

Surrealdb has this feature that allows in-memory instances to be created, this is achieved by using the `mem://` URL. 
The problem is that the public interface to create a broker instance is the url, which means if we create the in-memory instance somewhere else (the code that uses the broker), there is no guarantee (at least the documentation does not mention that) creating a new surrealdb `mem://` instance will reuse the existing one and not create a separate instance.

Added a way to pass the instance directly, and prioritise using that instance over creating a new connection. Don't think there is a better way than feature-specific fields, definitely open to suggestions.

**Comment by densumesh** [comment]
this looks good just needs a rebase onto main and then we can merge!

**Comment by danigiri** [comment]
have done extensive testing with a real world scenario and found some subtle issues (race conditions, etc.), working on multiple fixes, so will move to draft in the meantime

**Comment by danigiri** [comment]
tests pass without locks

**Comment by danigiri** [comment]
cc @densumesh 

OK, addressed multiple issues when testing with realistic workloads
- surrealdb does not really like individual record fetches (see https://github.com/surrealdb/surrealdb/issues/6086)
- mitigation is to have a `try_consume_batch` public method
- the combination of the above and `live()` consumption when there is no data means performance on surrealdb is quite adequate on a laptop, and can easily process 10's of thousands of messages quite quickly (queues are most of the time at zero)
- updated the documentation and removed some code duplication
- there are some smart retries when there are recoverable transactions
-  ran `cargo clippy --features surrealdb --fix --allow-dirty` again

After the above fixes, performance is good on a real world usage scenario, and there are no surfaced transactional issues. Given that, opening up the PR again.

Very happy to discuss `try_consume_batch`, use different naming, etc. 

Thanks in advance!

Ran tests again on a local surrealdb instance:
```
     Running unittests src/lib.rs (target/release/deps/broccoli_queue-6108306b2014ce3f)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/edge_cases.rs (target/release/deps/edge_cases-20001cb52353eeff)

running 8 tests
test test_concurrent_consume ... ok
test test_empty_payload ... ok
test test_invalid_broker_url ... ok
test test_message_ordering ... ok
test test_multiple_batch_publish_and_consume ... ok
test test_multiple_batch_publish_and_handler ... ok
test test_ttl_not_implemented ... ok
test test_very_large_payload ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.62s

     Running tests/fairness.rs (target/release/deps/fairness-a0e954979e704bee)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/happy_path.rs (target/release/deps/happy_path-73aa7d16c133a421)

running 12 tests
test test_batch_publish_and_consume ... ok
test test_delayed_message ... ok
test test_message_acknowledgment ... ok
test test_message_auto_ack ... ok
test test_message_cancellation ... ok
test test_message_priority ... ok
test test_message_retry ... ok
test test_process_messages ... ok
test test_process_messages_with_handlers ... ok
test test_publish_and_consume ... ok
test test_scheduled_message ... ok
test test_try_consume_batch ... ok

test result: ok. 12 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 6.45s

     Running tests/management.rs (target/release/deps/management-154ee5aba3de25fe)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests broccoli_queue

running 2 tests
test src/queue.rs - queue::BroccoliQueue::process_messages (line 701) - compile ... ok
test src/queue.rs - queue::BroccoliQueue::process_messages_with_handlers (line 849) - compile ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.23s
```



Latest benchmark figures

Screenshot 2025-07-01 at 11 22 33

**Comment by densumesh** [comment]
seems like we have a couple rebase errors, need to rebase on the main branch, and then we can merge. 
image

**Comment by danigiri** [comment]
### changes
added a consumer configuration  called `handler_ack`.  If we use `process_message`or `process_messages_with_handlers` we used to always ack messages. This prevents use cases like:
- having the handler spawn a long-running task and returning immediately
- having some other type of reaper process that is untied
If this is set to false, successful Ok()'s from the callback fn's do not trigger an acknowledge and it's the caller's responsibility to ACK as needed.
Note that handler errors still always trigger a reject.


### clippy
ran it on latest version, made some small changes, it did spit some compilation errors, assuming they are false positives

### tests

```
     Running unittests src/lib.rs (target/release/deps/broccoli_queue-cb687fcc89d05ff9)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

     Running tests/edge_cases.rs (target/release/deps/edge_cases-b7892b1c580d9924)

running 8 tests
test test_concurrent_consume ... ok
test test_empty_payload ... ok
test test_invalid_broker_url ... ok
test test_message_ordering ... ok
test test_multiple_batch_publish_and_consume ... ok
test test_multiple_batch_publish_and_handler ... ok
test test_ttl_not_implemented ... ok
test test_very_large_payload ... ok

test result: ok. 8 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 2.61s

     Running tests/fairness.rs (target/release/deps/fairness-96b314741ef14cb6)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

 Running tests/happy_path.rs (target/release/deps/happy_path-737e9bc30ee0b476)

running 12 tests
test test_batch_publish_and_consume ... ok
test test_delayed_message ... ok
test test_message_acknowledgment ... ok
test test_message_auto_ack ... ok
test test_message_cancellation ... ok
test test_message_priority ... ok
test test_message_retry ... ok
test test_process_messages ... ok
test test_process_messages_with_handlers ... ok
test test_publish_and_consume ... ok
test test_scheduled_message ... ok
test test_try_consume_batch ... ok

test result: ok. 12 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 6.40s

     Running tests/management.rs (target/release/deps/management-270175c472ff0622)

running 0 tests

test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

   Doc-tests broccoli_queue

running 2 tests
test src/queue.rs - queue::BroccoliQueue::process_messages (line 717) - compile ... ok
test src/queue.rs - queue::BroccoliQueue::process_messages_with_handlers (line 870) - compile ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.26s
```

**Comment by danigiri** [comment]
@densumesh thanks again for the review, merged with upstream and ran clippy.

From internal real-world testing, proposed a config change around auto-ack, see PR desc for details.

Thanks in advance!

**Review by densumesh — APPROVED** [review]

**Review by densumesh — APPROVED** [review]

File: src/brokers/surrealdb/broker.rs (hunk 4 above)
Comment (danigiri) [reply src/brokers/surrealdb/broker.rs:237]: Removed mutex completely, relying on the database for consistency, there is an optimistic locking model. Best results have been with a single consumer per topic.

**Comment by danigiri** [comment]
Thanks for the review and the merge @densumesh !
I am testing this with real world data and usage, will spin new PRs as needed.
